clang/lib/Driver/Driver.cpp

```diff
@@ -979,6 +1000,45 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
     return;
   }
 
+  //
+  // SYCL
+  //
+  // We need to generate a SYCL toolchain if the user specified -fsycl.
+  // If -fsycl is supplied without any of these we will assume SPIR-V.
+  bool HasValidSYCLRuntime =
+      C.getInputArgs().hasFlag(options::OPT_fsycl, options::OPT_fno_sycl,
+                               false);
+
+  auto argSYCLIncompatible = [&](OptSpecifier OptId) {
+    if (!HasValidSYCLRuntime)
+      return;
+    if (Arg *IncompatArg = C.getInputArgs().getLastArg(OptId))
+      Diag(clang::diag::err_drv_argument_not_allowed_with)
+          << IncompatArg->getSpelling() << "-fsycl";
+  };
+  // -static-libstdc++ is not compatible with -fsycl.
+  argSYCLIncompatible(options::OPT_static_libstdcxx);
+  // -ffreestanding cannot be used with -fsycl
+  argSYCLIncompatible(options::OPT_ffreestanding);
```

Comment:
Using `HasValidSYCLRuntime` name for the `hasFlag(OPT_fsycl)` return value is fine only for these few lines. These are checking the compatibility with the flags which configure different runtime settings.

But later code looks a bit strange. We probably should use something like `IsSYCL` as HIP/CUDA does here: https://github.com/mdtoguchi/llvm-project/blob/d819f8911ad6360b667289bd90eb5b6401105335/clang/lib/Driver/Driver.cpp#L799-L809.

Reply:
Updated to use `IsSYCL`